### PR TITLE
GROOVY-10034: do not cast non-primitive array when target is Object[]

### PR DESCRIPTION
**Pull request by eric-milles**
https://issues.apache.org/jira/browse/GROOVY-10034

### CODE REVIEW

File: src/main/java/org/codehaus/groovy/classgen/asm/OperandStack.java [hunk 1]
```diff
@@ -338,6 +338,12 @@ private void doConvertAndCast(ClassNode targetType, final boolean coerce) {
             return;
         }
 
+        // GROOVY-10034: no need to cast non-primitive array when target is "java.lang.Object[]"
```
Comment (danielsun1106) [inline src/main/java/org/codehaus/groovy/classgen/asm/OperandStack.java:341]: How about making the check more general? If the `top` component type is inherited from `target` component type, no need to cast type. The following code could be tweaked as:
```java
        if (targetType.isArray() && !ClassHelper.isPrimitiveType(targetType.getComponentType()) 
                  && top.isArray() && !ClassHelper.isPrimitiveType(top.getComponentType())
                  && targetType.getComponentType().isAssignableFrom(top.getComponentType())) {
            return;
        }
```

Comment (danielsun1106) [reply src/main/java/org/codehaus/groovy/classgen/asm/OperandStack.java:341]: It seems that we could make the check simpler, we do not add extra checks but tweak the existing check:

https://github.com/apache/groovy/blob/master/src/main/java/org/codehaus/groovy/classgen/asm/OperandStack.java#L334
```java
if (targetType == top) return;
```
tweaked as:
```java
if (targetType.getTypeClass().isAssignableFrom(top.getTypeClass())) return;
```

Comment (eric-milles) [reply src/main/java/org/codehaus/groovy/classgen/asm/OperandStack.java:341]: I wanted to keep the scope limited.  This is so widely used.  You could certainly add a new issue that references this one and I could take a look at it after I get through with array stuff.

Comment (eric-milles) [reply src/main/java/org/codehaus/groovy/classgen/asm/OperandStack.java:341]: `getTypeClass()` is a known unsafe method to use.  I would caution against that without thorough testing.

Comment (danielsun1106) [reply src/main/java/org/codehaus/groovy/classgen/asm/OperandStack.java:341]: OK. 
Here is the JIRA ticket to track the discusstion: https://issues.apache.org/jira/browse/GROOVY-10035

**Comment by danielsun1106** [comment]
Merged. Thanks!

**Comment by danielsun1106** [comment]
I created a PR https://github.com/apache/groovy/pull/1561 based on your current PR to tweak the check a bit.

**Pull request merged**